go/beacon/api/api.go

```diff
@@ -15,41 +16,152 @@ const (
+	// messages on epoch transitions. If an epoch transition happens
+	// before the previous epoch is read from the channel, the old
+	// epochs are overwritten.
+	//
+	// Upon subscription the current epoch is sent immediately.
+	WatchLatestEpoch() (<-chan EpochTime, *pubsub.Subscription)
+
+	// GetBeacon gets the beacon for the provided block height.
+	// Calling this method with height `0` should return the
+	// beacon for the latest finalized block.
 	GetBeacon(context.Context, int64) ([]byte, error)
 
 	// StateToGenesis returns the genesis state at specified block height.
 	StateToGenesis(context.Context, int64) (*Genesis, error)
 }
 
-// Genesis is the beacon genesis state.
+// SetableBackend is a Backend that supports setting the current epoch.
+type SetableBackend interface {
+	Backend
+
+	// SetEpoch sets the current epoch.
+	SetEpoch(context.Context, EpochTime) error
+}
+
+// Genesis is the genesis state.
 type Genesis struct {
+	// Base is the starting epoch.
+	Base EpochTime `json:"base"`
+
 	// Parameters are the beacon consensus parameters.
 	Parameters ConsensusParameters `json:"params"`
 }
 
 // ConsensusParameters are the beacon consensus parameters.
 type ConsensusParameters struct {
+	// Backend is the beacon backend.
+	Backend string `json:"backend"`
+
+	// DebugMockBackend is flag for enabling the mock epochtime backend.
+	DebugMockBackend bool `json:"debug_mock_backend,omitempty"`
+
 	// DebugDeterministic is true iff the output should be deterministic.
 	DebugDeterministic bool `json:"debug_deterministic,omitempty"`
+
+	// InsecureParameters are the beacon parameters for the insecure backend.
+	InsecureParameters *InsecureParameters `json:"insecure_parameters,omitempty"`
+
+	// SCRAPEParameters are the beacon parameters for the SCRAPE backend.
+	SCRAPEParameters *SCRAPEParameters `json:"scrape_parameters,omitempty"`
+}
+
+// InsecureParameters are the beacon parameters for the insecure backend.
+type InsecureParameters struct {
+	// Interval is the epoch interval (in blocks).
+	Interval int64 `json:"interval,omitempty"`
 }
 
 // SanityCheck does basic sanity checking on the genesis state.
 func (g *Genesis) SanityCheck() error {
-	unsafeFlags := g.Parameters.DebugDeterministic
+	switch g.Parameters.Backend {
+	case BackendInsecure:
+		params := g.Parameters.InsecureParameters
+		if params == nil {
+			return fmt.Errorf("beacon: sanity check failed: insecure backend not configured")
+		}
+
+		if params.Interval <= 0 && !g.Parameters.DebugMockBackend {
+			return fmt.Errorf("beacon: sanity check failed: epoch interval must be > 0")
+		}
+	case BackendSCRAPE:
+		params := g.Parameters.SCRAPEParameters
+		if params == nil {
+			return fmt.Errorf("beacon: sanity check failed: SCRAPE backend not configured")
+		}
+
+		if params.Participants <= 1 {
+			return fmt.Errorf("beacon: sanity check failed: SCRAPE participants must be > 1")
+		}
+		if n := params.Threshold; n <= 1 || n > params.Participants {
```

Comment:
(Using $\LaTeX$ style to refer to stuff from the paper and `this` to refer to code.)

In the SCRAPE paper, the thresholds used appears to be $t=n/2$ or `params.Participants/2`.  To make auditing easier, I suggest renaming the `n` here (and/or below at L140) to be `t`, and please add comments why we would want to allow the full range of possible values.  Is this threshold $|\mathcal{C}|$?  It appears to be from the comment in `go/common/crypto/scrape/scrape.go`, where it says this threshold probably should be at least `params.PVSSThreshold`, so this should probably be sanityy checked here, unless we don't want to enforce it for some reason (please provide reason somewhere).  The `New` there does not enforce this either.

Is $n$ or `params.Participants` expected to be about the size of the validator committee?

Reply:
> Is $n$ or `params.Participants` expected to be about the size of the validator committee?

Depends on how big the validator committee is, usually this would be a subset of the committee due to performance reasons. E.g. at most ~20.

Reply:
Truth be told, unless there's a strong reason not to do so, I'm tempted to unify Threshold and PVSSThreshold since I have a hard time envisioning a world where they are set to different values.

Reply:
Regarding the allowable values, as far as my understanding of the protocol goes, like all of these sort of commit/reveal type schemes, the output is unbiased as long as 1 participant is honest.  Naturally we will likely use something that is more than 1 in practice, but if people wish to do something less sensible that's up to them.

Reply:
I agree that unifying the two thresholds could be reasonable.  Generally the goal is to be able to withstand a large number of dishonest nodes while allowing some honest node failures (or eclipsing).  If `PVSSThreshold` were too low then it means if there are that many dishonest nodes they can reconstruct votes of other participants in private and choosing their own $s_j$ adaptively to influence $\rho$.  So the number of dishonest nodes has to be less than `PVSSThreshold`, and the number of honest but unavailable nodes has to be less than `n - Threshold`.

I think this is accurate:

```
|---------------------|------------|-----------|-------------------|
0                                                                  n
|<- dishonest ->      |     <- honest ->                           |
| PVSSThreshold                    |                               |
|   available / Threshold                      |     down          |
```
while |dishonest| < `Threshold` < `PVSSThreshold` is feasible, it doesn't make sense since if `PVSSThreshold` nodes are available to run Reconstruction (to publish $\tilde{s}^a_j$) then they will have posted their shares with $\textsf{Open}(s^j,r_j)$ and PVSS will never need to be used.